pkg/dns/treecache.go

```diff
@@ -0,0 +1,312 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+const (
+	dataFile = "data.dat"
+	crcFile  = "data.crc"
+)
+
+type object interface{}
+
+type TreeCache struct {
+	ChildNodes map[string]*TreeCache
+	Entries    map[string]interface{}
+	m          *sync.RWMutex
+}
+
+func NewTreeCache() *TreeCache {
+	return &TreeCache{
+		ChildNodes: make(map[string]*TreeCache),
+		Entries:    make(map[string]interface{}),
+		m:          &sync.RWMutex{},
+	}
+}
+
+func Deserialize(dir string) (*TreeCache, error) {
+	b, err := ioutil.ReadFile(path.Join(dir, dataFile))
+	if err != nil {
+		return nil, err
+	}
+	var hash []byte
+	hash, err = ioutil.ReadFile(path.Join(dir, crcFile))
+	if err != nil {
+		return nil, err
+	}
+	if !reflect.DeepEqual(hash, getMD5(b)) {
```

Comment:
Just [`bytes.Equal`](https://golang.org/pkg/bytes/#Equal) is sufficient? `reflect.DeepEqual()` seems both heavy-weight and makes the code less clear. I had to go and double check the type of `hash` and the return type of `getMD5()`.